### PR TITLE
Use a lister for Prowjobs in sinker

### PR DESCRIPTION
**Pull request by alvaroaleman**
This PR changes sinker to use a lister for Prowjobs in order to not do a full LIST of all prowjobs every sync iteration. It also enabled leader election, as that is now easy to do.

Some stats, just from `kubectl top pods`, unfortunately there is no Prometheus yet, 2.5k Prowjobs, 5s `resync_period`, nothing going on at all:

* CPU now stays at constantly 15m -20m, was 1-60M (This currently has a huge variance, I presume this will spike very high when doing the deserialization but thats only visible in a very short timeframe)
* Mem now stays at ~110Mi, was ~80Mi
* Sync duration is now at 200-300MS, was 2-3S, presumably because ~no api call is involved anymore~ the 2.5k Prowjob List call takes very long

### CODE REVIEW

File: prow/cluster/sinker_rbac.yaml [hunk 1]
```diff
@@ -17,6 +17,24 @@ rules:
...
       - delete
       - list
+      - watch
+      - get
+  - apiGroups:
+      - ""
+    resources:
+      - configmaps
+    resourceNames:
+      - prow-sinker-leaderlock
```
Comment (stevekuznetsov) [inline prow/cluster/sinker_rbac.yaml]: Why do we need this if we only have one replica of `sinker`?

Comment (alvaroaleman) [reply prow/cluster/sinker_rbac.yaml]: Upgrades

Comment (alvaroaleman) [reply prow/cluster/sinker_rbac.yaml]: If you think that warrants a separate discussion I will split it off this PR

**Comment by stevekuznetsov** [comment]
No, upgrades make sense -- let's rebase this one and I think it's good

/lgtm
/approve

**Comment by alvaroaleman** [comment]
@stevekuznetsov rebased

File: prow/cluster/sinker_deployment.yaml [hunk 2]
```diff
@@ -23,6 +23,12 @@ spec:
         - --config-path=/etc/config/config.yaml
         - --job-config-path=/etc/job-config
         - --dry-run=false
+        - --namespace=$(MY_POD_NAMESPACE)
+       env:
+       - name: MY_POD_NAMESPACE
+         valueFrom:
+          fieldRef:
+            fieldPath: metadata.namespace
```
Comment (cjwagner) [inline prow/cluster/sinker_deployment.yaml]: This isn't the `test-pods` namespace is it?

Comment (cjwagner) [reply prow/cluster/sinker_deployment.yaml]: Actually looks more like the env var name is misleading. Should be `MY_PROWJOB_NAMESPACE`?

We really should be consuming theses values from config though. We shouldn't assume that `sinker` is running in the same namespace the ProwJobs are in: https://github.com/kubernetes/test-infra/blob/036124a3a282bac8859218cf7112721f151b71c1/prow/config/config.go#L95-L104

Comment (alvaroaleman) [reply prow/cluster/sinker_deployment.yaml]: It's the namespace the Sinker Pod runs in, it's being used for the leader election only

**Pull request merged**